cmd/skaffold/app/cmd/config/util.go

```diff
@@ -189,16 +205,49 @@ func GetLocalCluster() (bool, error) {
 		// if no value is set for this cluster, fall back to the global setting
 		globalCfg, err := GetGlobalConfig()
 		if err != nil {
+			cachedLocalCluster = &localCluster
 			return localCluster, errors.Wrap(err, "retrieving global config")
 		}
 		if globalCfg != nil && globalCfg.LocalCluster != nil {
 			localCluster = *globalCfg.LocalCluster
 		}
 	}
 
+	cachedLocalCluster = &localCluster
 	return localCluster, nil
 }
 
+func GetInsecureRegistries() ([]string, error) {
+	if cachedInsecureRegistries != nil {
+		return *cachedInsecureRegistries, nil
+	}
+	cfg, err := GetConfigForKubectx()
+	registries := []string{}
```

Comment:
I don't think we want this to be nil. if we search through the configs and don't find a match (either in the context-specific one or the global one), then we don't have a value set, and we would *want* the cached value to be empty. if we kept it nil, the cached value would be `&(nil)`, which would panic.

Reply:
I wrote a snippet over [here](https://play.golang.org/p/uHChOEdL2So), 
```
package main

import "fmt"

func main() {
        entries := []string{}
        cache := &entries
        if cache != nil {
          fmt.Println("For an empty list -> cache is detected as not empty. It should be empty instead")
        } else {
	  fmt.Println("For an empty list -> cache is detected as empty correctly")
	}
}
```
For []string{}, cachedInsecureRegistries != nil return true. 
I see, runner.NewForConfig() calls `configutil.GetInsecureRegistries()`. If it is expected that runner will be created multiple times and we want to retry retrieving insecure registries if we could not fetch the gobal config or kubectl config first time, we shd be setting it to "nil".
May be i am missing something.

Reply:
I think your main issue might be with the way that this value was being cached? I removed the caching from this PR, so we could pick that conversation back up in the future when I reimplement it.

I'm not sure I understand your comment about multiple runners though, each skaffold run only ever creates one runner. Regardless, the values `[]string{}` and `nil` are effectively equivalent here: we're combining each of the lists of insecure registries together in the runner to get a full list of registries specified in all three places:

https://github.com/GoogleContainerTools/skaffold/blob/a3a739b554e329da99b90acdc6f5d9baf6ddb711/pkg/skaffold/runner/runner.go#L94-L104

Reply:
> I think your main issue might be with the way that this value was being cached? I removed the caching from this PR, so we could pick that conversation back up in the future when I reimplement it.

Correct, i dont think the way we have it now will work.

> I'm not sure I understand your comment about multiple runners though, each skaffold run only ever creates one runner. 

From my understanding of code, i feel there is just one runner per skaffold run. I also found just one usage for `GetInsecureRegistries`
```
~/go/src/github.com/GoogleContainerTools/skaffold: (pr/nkubala/1824)$ git grep 'GetInsecureRegistries'
cmd/skaffold/app/cmd/config/util.go:func GetInsecureRegistries() ([]string, error) {
pkg/skaffold/runner/runner.go:  cfgRegistries, err := configutil.GetInsecureRegistries()
```
And that makes me wonder why do we need to cache this value? Maybe i should have made it clear in first comment instead of speculating "are multiple runners expected?".

Reply:
> Correct, i dont think the way we have it now will work.

I think I'm confused :smile: could you explain why?

> And that makes me wonder why do we need to cache this value?

there's no inherent reason why, the reason I had added it earlier was mostly for the other values in the global config (e.g. `default-repo`) which are being retrieved multiple times throughout a skaffold run; I was digging around in there and noticed that so I added it, and figured I may as well cache all the values in the config even if they're only being retrieved once. shouldn't have been added in this PR though :+1: